File: widget/src/main/java/ru/kest/trainswidget/data/DataProviderImpl.java
package ru.kest.trainswidget.data;

import android.content.Context;
import android.location.Location;

import java.util.List;

import ru.kest.trainswidget.model.domain.NearestStation;
import ru.kest.trainswidget.model.domain.TrainThread;
import ru.kest.trainswidget.util.PreferencesUtil;

/**
 * Created by KKharitonov on 14.02.2016.
 */
public class DataProviderImpl implements DataProvider{

    private FieldDataStorage dataStorage;
    private Context context;

    protected DataProviderImpl(Context context) {
        this.context = context;
        this.dataStorage = PreferencesUtil.createDataStorage(context);
    }

    @Override
    public List<TrainThread> getTrainsFromHomeToWork() {
        return dataStorage.getTrainsFromHomeToWork();
    }

    @Override
    public void setTrainsFromHomeToWork(List<TrainThread> trainsFromHomeToWork) {
        PreferencesUtil.saveTrainsFromHomeToWork(context, trainsFromHomeToWork);
        dataStorage.setTrainsFromHomeToWork(trainsFromHomeToWork);
    }

    @Override
    public List<TrainThread> getTrainsFromWorkToHome() {
        return dataStorage.getTrainsFromWorkToHome();
    }

    @Override
    public void setTrainsFromWorkToHome(List<TrainThread> trainsFromWorkToHome) {
        PreferencesUtil.saveTrainsFromWorkToHome(context, trainsFromWorkToHome);
        dataStorage.setTrainsFromWorkToHome(trainsFromWorkToHome);
    }

    @Override
    public boolean isSetTrainThreads() {
        return getTrainsFromHomeToWork() != null && getTrainsFromWorkToHome() != null;
    }

    @Override
    public TrainThread getThreadByHash(int hash) {
        TrainThread result = getThreadByHash(getTrainsFromHomeToWork(), hash);
        if (result == null) {
            result = getThreadByHash(getTrainsFromWorkToHome(), hash);
        }
        return result;
    }

    @Override
    public Location getLastLocation() {
        return dataStorage.getLastLocation();
    }

    @Override
    public void setLastLocation(Location location) {
        PreferencesUtil.saveLastLocation(context, location);
        dataStorage.setLastLocation(location);
    }

    @Override
    public boolean isSetLastLocation() {
        return getLastLocation() != null;
    }

    @Override
    public TrainThread getNotificationTrain() {
        return dataStorage.getNotificationTrain();
    }

    @Override
    public void setNotificationTrain(TrainThread notificationTrain) {
        PreferencesUtil.saveNotificationTrain(context, notificationTrain);
        dataStorage.setNotificationTrain(notificationTrain);
    }

    @Override
    public boolean isSetNotificationTrain() {
        return getNotificationTrain() != null;
    }

    @Override
    public NearestStation getNearestStation() {
        if (isSetLastLocation()) {
            return NearestStation.getNearestStation(getLastLocation());
        }
        return null;
    }

    private TrainThread getThreadByHash(List<TrainThread> threads, int hash) {
        if (threads != null) {
            for (TrainThread thread : threads) {
                if (thread.hashCode() == hash) {
                    return thread;
                }
            }
        }
        return null;
    }
}

File: widget/src/main/java/ru/kest/trainswidget/data/DataProviderImpl.kt
package ru.kest.trainswidget.data

import android.content.Context
import android.location.Location
import android.util.Log
import ru.kest.trainswidget.LOG_TAG

import ru.kest.trainswidget.model.domain.NearestStation
import ru.kest.trainswidget.model.domain.TrainThread
import ru.kest.trainswidget.util.PreferencesUtil

/**
 * Имплементация хранилища данных на основе SharedPreferences
 *
 * Created by KKharitonov on 14.02.2016.
 */
class DataProviderImpl(private val context: Context) : DataProvider {

    init {
        Log.d(LOG_TAG, "## Read properties")
    }

    private val dataStorage: FieldDataStorage = PreferencesUtil.createDataStorage(context)

    override var trainsFromHomeToWork: List<TrainThread>
        get() = dataStorage.trainsFromHomeToWork
        set(value) {
            PreferencesUtil.saveTrainsFromHomeToWork(context, value)
            dataStorage.trainsFromHomeToWork = value
        }

    override var trainsFromWorkToHome: List<TrainThread>
        get() = dataStorage.trainsFromWorkToHome
        set(value) {
            PreferencesUtil.saveTrainsFromWorkToHome(context, value)
            dataStorage.trainsFromWorkToHome = value
        }

    override val isSetTrainThreads: Boolean
        get() = trainsFromHomeToWork.isNotEmpty() && trainsFromWorkToHome.isNotEmpty()

    override var lastLocation: Location?
        get() = dataStorage.lastLocation
        set(value) {
            if (value != null) {
                PreferencesUtil.saveLastLocation(context, value)
            }
            dataStorage.lastLocation = value
        }

    override val isSetLastLocation: Boolean
        get() = lastLocation != null

    override var notificationTrain: TrainThread?
        get() = dataStorage.notificationTrain
        set(value) {
            if (value != null) {
                PreferencesUtil.saveNotificationTrain(context, value)
            }
            dataStorage.notificationTrain = value
        }

    override val isSetNotificationTrain: Boolean
        get() = notificationTrain != null

    override val nearestStation: NearestStation?
        get() = if (lastLocation != null) NearestStation.getNearestStation(lastLocation!!) else null


    override fun getThreadByHash(hash: Int): TrainThread? {
        var result = getThreadByHash(trainsFromHomeToWork, hash)
        if (result == null) {
            result = getThreadByHash(trainsFromWorkToHome, hash)
        }
        return result
    }

    private fun getThreadByHash(threads: List<TrainThread>?, hash: Int): TrainThread? {
        if (threads != null) {
            @Suppress("LoopToCallChain")
            for (thread in threads) {
                if (thread.hashCode() == hash) {
                    return thread
                }
            }
        }
        return null
    }
}
